src/azure-cli-core/azure/cli/core/commands/progress.py

```diff
@@ -0,0 +1,180 @@
+class ProgressType(Enum):  # pylint: disable=too-few-public-methods
+    """ the types of progress """
+    Determinate = 0
+    Indeterminate = 1
+    # for views that want to accept both determinate and indeterminate progres
+    Both = 2
+
+
+class _ProgressViewBase(object):
+    """ a view base for progress reporting """
+    def __init__(self, out, progress_type, format_percent=None):
+        self.out = out
+        assert isinstance(progress_type, ProgressType)
+        self.progress_type = progress_type
+        self.format_percent = format_percent
+
+    def write(self, args):
+        """ writes the progress """
+        pass
+
+    def flush(self):
+        """ flushes the message out the pipeline"""
+        self.out.flush()
+
+    def get_type(self):
+        """ returns the progress type """
+        return self.progress_type
+
+
+class DeterminateProgressView(_ProgressViewBase):
+    """ a view base for progress reporting """
+    def __init__(self, out, format_percent=None):
+        super(DeterminateProgressView, self).__init__(out, ProgressType.Determinate, format_percent)
+
+
+class InDeterminateProgressView(_ProgressViewBase):
+    """ a view base for progress reporting """
+    def __init__(self, out, format_percent=None):
+        super(InDeterminateProgressView, self).__init__(
+            out, ProgressType.Indeterminate, format_percent)
+
+
+class DetProgressReporter(object):
+    """ generic progress reporter """
+    def __init__(self, total_value=1):
+        self.message = ''
+        self.curr_val = 0 if total_value else None
+        self.total_val = total_value
+
+    def add(self, kwargs):
+        """
+        adds a progress report
+        :param kwargs: dictionary containing 'message', 'total_val', 'value'
+        """
+        message = kwargs.get('message', '')
+        total_val = kwargs.get('total_val', 1.0)
+        value = kwargs.get('value', 0.0)
+        assert value >= 0 and value <= total_val and total_val >= 0
+        self.total_val = total_val
+        self.curr_val = value
+        self.message = message
+
+    def report(self):
+        """ report the progress """
+        percent = self.curr_val / self.total_val
+        return {'message': self.message, 'percent': percent}
+
+
+class InDetProgressReporter(object):
+    """ generic progress reporter """
+    def __init__(self):
+        self.message = ''
+
+    def add(self, kwargs):
+        """ adds a progress report """
+        self.message = kwargs.get('message', '')
+
+    def report(self):
+        """ report the progress """
+        return {'message': self.message}
+
+
+class ProgressHook(object):
+    """ sends the progress to the view """
+    def __init__(self, progress_type):
+        if progress_type == ProgressType.Determinate:
+            self.reporter = DetProgressReporter()
+        elif progress_type == ProgressType.Indeterminate:
+            self.reporter = InDetProgressReporter()  # pylint: disable=redefined-variable-type
```

Comment:
`If-Else` here is anti-pattern. What should be done is let the caller of the `__init__` to determine what type of the reporter and send in instance.